### PR TITLE
Update llvm backends to use llvm 15

### LINKED ISSUE #1116: Update llvm backends to use opaque pointers
As of LLVM 16,  C API no longer has [`LLVMConstGEP`](https://github.com/ballerina-platform/nballerina/blob/b349ae2f4540b260650271f7891403dae012c370/llvm_jni/modules/jni.llvm/Context.bal#L125) and [`LLVMConstInBoundsGEP`](https://github.com/ballerina-platform/nballerina/blob/b349ae2f4540b260650271f7891403dae012c370/llvm_jni/modules/jni.llvm/Context.bal#L131) functions, instead, it has [`LLVMConstGEP2`](https://llvm.org/doxygen/group__LLVMCCoreValueConstantExpressions.html#gae355c965dacae2f840ad9c1e8b60f885) and [`LLVMConstInBoundsGEP2`](https://llvm.org/doxygen/group__LLVMCCoreValueConstantExpressions.html#ga7bd51db0c1eec33d38b10cccc356d35f). Once [bytedeco llvm presets](https://github.com/bytedeco/javacpp-presets/tree/master/llvm) has been updated to llvm 16 we will need to rewrite those functions in `jni.llvm`

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Update llvm backends to use llvm 15. (The main change is the switch to opaque pointers)
Resolves #1116

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/nback/number.bal [hunk 1]
```diff
@@ -305,9 +305,9 @@ function buildDecimalNegate(llvm:Builder builder, Scaffold scaffold, bir:Decimal
 }
 
 final readonly & map<llvm:IntrinsicFunctionName> binaryIntIntrinsics = {
-    "+": "sadd.with.overflow.i64",
-    "-": "ssub.with.overflow.i64",
-    "*": "smul.with.overflow.i64"
+    "+": "sadd.with.overflow.i64.i64",
```
Comment (jclark) [inline compiler/modules/nback/number.bal]: Is this available in LLVM 13?

**Comment by jclark** [comment]
Mention that this is also upgrading to LLVM 15.

File: compiler/modules/nback/number.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/number.bal]: Yes. Created #1150 with just this change

File: compiler/modules/print.llvm/llvm.bal [hunk 2]
```diff
@@ -1610,6 +1598,31 @@ function typeToString(RetType ty, Context context, boolean forceInline=false) re
     }
 }
 
+// use to mark a named struct type as used (so we include that type in output) when it is referenced via a pointer
+function markNamedTypeAsUsed(RetType ty, Context context) {
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: Name feels a bit long. How about `useNamedType` or `includeNamedType`?

**Review by jclark — DISMISSED** [review]

File: compiler/modules/print.llvm/llvm.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: fixed

**Review by jclark — APPROVED** [review]

**Pull request merged**